data/core/macros/items.cfg

```diff
@@ -261,11 +261,22 @@
 #enddef
 
 #define OBJ_POTION_HOLY X Y ID
-    {PICKUPPABLE_ITEM {ID} {X} {Y} side=1 items/holy-water.png
+    {PICKUPPABLE_ITEM {ID} {X} {Y} (
+        side=1
+        [has_attack]
+            range=melee
+        [/has_attack]
+        [not]
+            [has_attack]
+                range=melee
+                type=arcane
+            [/has_attack]
+        [/not]
+    ) items/holy-water.png
```

Comment:
OK. Tested.

Dark Adept: no (no melee at all)
Moremiru: no (only melee weapon is already arcane)
Paladin: yes, once (one melee weapon is not arcane, one is already)
Knight: yes, once (two weapons, both melee not arcane, both become arcane)
Konrad: yes, once (only weapon is not arcane)

I'll push the change once I can clear the mess I made testing this in the midst of debugging Li'sar and her initiative ability causing a unit to be clicked at scenario-start which is a pain in the arse because you have to unclick it since you can't move it, then click your unit.